### PR TITLE
Migrate application_package to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Handle a null product ID in `BuildableIOSApp.fromProject()`, and a test to validate.

Part of #71511

### CODE REVIEW

File: packages/flutter_tools/lib/src/ios/application_package.dart [hunk 1]
```diff
@@ -106,26 +101,26 @@ abstract class IOSApp extends ApplicationPackage {
...
-  Directory get appDeltaDirectory;
+  Directory? get appDeltaDirectory;
 }
 
 class BuildableIOSApp extends IOSApp {
-  BuildableIOSApp(this.project, String projectBundleId, String hostAppBundleName)
+  BuildableIOSApp(this.project, String? projectBundleId, String? hostAppBundleName)
     : _hostAppBundleName = hostAppBundleName,
-      super(projectBundleId: projectBundleId);
+      super(projectBundleId: projectBundleId ?? 'com.example.Runner');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/ios/application_package.dart:110]: What are the mechanics of filling in `com.example.Runner'` here? If this field was actually missing are we going to fail somewhere else? should we just fail earlier?

**Comment by jonahwilliams** [comment]
Friendly ping :)

File: packages/flutter_tools/lib/src/ios/application_package.dart (hunk 1 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/ios/application_package.dart:110]: I have a better idea.  `projectBundleId` can be made nonnull if `BuildableIOSApp.fromProject` can return a nullable project, and its only callsite `IOSApp.fromIosProject` handles a null project.

That probably didn't make sense, diff (with test) incoming...

Comment (jmagman) [reply packages/flutter_tools/lib/src/ios/application_package.dart]: See https://github.com/flutter/flutter/pull/83956/commits/609c217da0170518a6409358b1d8f304edeaf239

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**